File: docs/papers211_215_sequence/qa/p211_round2_root_reception01/EXECUTED_CAPTURE_SOURCE.py
#!/usr/bin/python3.10
"""Root's explicit complete-source documentary reuse, never a recorder run.

The 1138-line receiver was fully read. This capture follows the accepted
Round1 recheck mechanism; source, whole original output, package membership,
all full rich input keys and actual raw comparison are bound independently.
All new outputs are exclusive in this new root directory. ENV4 is literal.
"""
from hashlib import sha256
import json
import os
from pathlib import Path
import re
import stat
import subprocess
import time

ROOT=Path('/root/autodl-tmp/symbolic_dynamics')
QA=ROOT/'docs/papers211_215_sequence/qa'
HERE=QA/'p211_round2_root_reception01'
A=QA/'p211_round2_physical_independent01'
SOURCE=A/'receive_physical05.py'
ENV={'PATH':'/usr/bin:/bin','LANG':'C.UTF-8','LC_ALL':'C.UTF-8','TZ':'UTC'}
READS={}
CHECKS=0
def need(x,label):
    global CHECKS
    CHECKS+=1
    if not x:
        raise AssertionError(label)
def pin(raw):
    return {'bytes':len(raw),'sha256':sha256(raw).hexdigest()}
def meta(s):
    return {k:getattr(s,'st_'+v) for k,v in (
        ('mode','mode'),('device','dev'),('inode','ino'),('uid','uid'),('gid','gid'),
        ('nlink','nlink'),('size','size'),('mtime_ns','mtime_ns'),('ctime_ns','ctime_ns'))}
def rich(p):
    p=Path(p)
    ls,st=p.lstat(),p.stat()
    need(stat.S_ISREG(st.st_mode),('ordinary resolved file',str(p)))
    resolved=str(p.resolve(strict=True))
    link=os.readlink(p) if stat.S_ISLNK(ls.st_mode) else None
    if p.is_relative_to(ROOT):
        need(resolved==str(p) and link is None,('ordinary workspace file',str(p)))
    raw=p.read_bytes()
    key={**pin(raw),'resolved':resolved,'symlink':link,'stat':meta(st),'lstat':meta(ls)}
    need(meta(p.lstat())==key['lstat'] and meta(p.stat())==key['stat'] and
         str(p.resolve(strict=True))==resolved and len(raw)==st.st_size,'whole rich read stable')
    need(str(p) not in READS or READS[str(p)]==key,('repeated original full key',str(p)))
    READS[str(p)]=key
    return raw,key
def read(p):
    return rich(p)[0]
def put(name,v):
    raw=v if isinstance(v,bytes) else (json.dumps(v,sort_keys=True,indent=2)+'\n').encode()
    with (HERE/name).open('xb') as f:
        f.write(raw)
        f.flush()
        os.fsync(f.fileno())
def package():
    raw=read(A/'SHA256SUMS')
    need(pin(raw)=={'bytes':6880,'sha256':'f1eac068f6f8dea79ae9cda7028b671a2d5e792af5836c2f15f84b0a09c82e39'},'entire final independent seal')
    rows={}
    need(raw.endswith(b'\n'),'full manifest LF')
    for line in raw.decode().splitlines():
        m=re.fullmatch(r'([0-9a-f]{64})  (.+)',line)
        need(m is not None,'manifest syntax')
        h,n=m.groups()
        need(n not in rows and n!='SHA256SUMS' and not Path(n).is_absolute() and
             all(x not in ('','.','..') for x in n.split('/')),'exact safe nonself membership')
        need(pin(read(A/n))['sha256']==h,('all independent payloads',n))
        rows[n]=h
    names=set()
    for p in A.rglob('*'):
        need(not p.is_symlink() and (p.is_dir() or p.is_file()),'ordinary complete independent tree')
        if p.is_file():
            names.add(str(p.relative_to(A)))
    need(len(rows)==75 and names==set(rows)|{'SHA256SUMS'},'all 75 payloads / 76 files')
    return raw
def command(label,argv,timeout):
    executable=rich(Path(argv[0]))[1]
    a={'argv':argv,'cwd':str(ROOT),'environment':ENV,'stdin':'subprocess.DEVNULL',
       'timeout_seconds':timeout,'started_epoch':time.time(),'executable_full_key':executable}
    put(label+'.ATTEMPT.json',a)
    error,code,state=None,None,'captured'
    try:
        r=subprocess.run(argv,cwd=ROOT,env=ENV,stdin=subprocess.DEVNULL,capture_output=True,timeout=timeout)
        out,err,code=r.stdout,r.stderr,r.returncode
    except subprocess.TimeoutExpired as e:
        out,err=e.stdout or b'',e.stderr or b''
        error={'type':type(e).__name__,'message':str(e)}
        state='partial_at_timeout; no exit invented'
    except OSError as e:
        out,err=b'',b''
        error={'type':type(e).__name__,'message':str(e)}
        state='launch_failed_or_unknown; no exit invented'
    put(label+'.stdout.raw',out)
    put(label+'.stderr.raw',err)
    n={**a,'ended_epoch':time.time(),'native_exit_code':code,'exception':error,'stream_capture_status':state,
       'stdout':pin(out),'stderr':pin(err)}
    put(label+'.NATIVE.json',n)
    need(error is None and type(code) is int and code==0 and err==b'',('actual complete native success',label))
    need(rich(Path(argv[0]))[1]==executable,'whole native tool unchanged')
    return out,n

need(Path.cwd()==ROOT and dict(os.environ)==ENV,'actual literal cwd and ENV4; no ambient collection')
read(Path(__file__).absolute())
read(HERE/'AUTHORITY.md')
put('EXECUTED_CAPTURE_SOURCE.py',read(Path(__file__).absolute()))
seal=package()
source=read(SOURCE)
need(pin(source)=={'bytes':77479,'sha256':'ad8bd8890f9d5415d2a0499485415cb42858eca9c48b04cdecaf7b9c4c0c0e9d'},'exact fully read receiver source')
raw=read(A/'run05/CHECKS.stdout.raw')
need(pin(raw)=={'bytes':3430632,'sha256':'ff715a36ffd80ada101e64bf39085606155d7ea3c87beefe80555910fee84fdc'},'entire actual independent output')
whole=json.loads(raw)
need(whole['checks']==488474 and whole['receiver_read_paths']==4821 and len(whole['READ_INPUTS'])==4821,'entire original physical input key')
allowed=set(whole['READ_INPUTS'])
for p,v in whole['READ_INPUTS'].items():
    need(rich(Path(p))[1]==v,('all accepted original rich rows before execution',p))
put('RECHECK_INPUTS_BEFORE.json',READS)
out,native=command('01_recheck',['/usr/bin/python3.10','-I','-S','-B',str(SOURCE)],180)
need(out==raw and json.loads(out)==whole,'entire original and recheck raw bytes and JSON identical')
compared,cmpnative=command('02_compare',['/usr/bin/cmp','--',str(A/'run05/CHECKS.stdout.raw'),str(HERE/'01_recheck.stdout.raw')],60)
need(compared==b'','actual native whole raw comparison')
for p,v in whole['READ_INPUTS'].items():
    need(rich(Path(p))[1]==v,('all accepted original rich rows after execution',p))
need(package()==seal and read(SOURCE)==source,'whole original seal/source unchanged after recheck')
for p,v in dict(READS).items():
    need(rich(Path(p))[1]==v,('complete root capture original key closes',p))
put('RECHECK_INPUTS_AFTER.json',READS)
result={'status':'PASS_ROOT_EXPLICIT_COMPLETE_PHYSICAL_RECEIVER_RECHECK_PENDING_ORIGINAL_RECEPTION',
        'root_capture_checks':CHECKS,'root_capture_read_paths':len(READS),'reused_receiver_checks':488474,
        'reused_receiver_read_paths':4821,'independent_payloads':75,'independent_files':76,
        'whole_raw_output':pin(out),'native_cmp_exit':cmpnative['native_exit_code'],
        'source':pin(source),'independent_seal':pin(seal),'new_science':0,'new_builds':0,'new_page_views':0,
        'paper_complete':False,'new_independent_design':False,'terminal_authority':False,
        'external':'OWNER_AMBER / HOLD_EXTERNAL'}
put('RECHECK_RESULT.json',result)
print(json.dumps(result,sort_keys=True))

File: symbolic_dynamics/docs/papers211_215_sequence/qa/p211_terminal_binding_source_root01/EXECUTED_CAPTURE_SOURCE.js
'use strict';
// Root receipt of source-only preparation and independent source metadata.
// Complete unchanged metadata helper is explicitly run, not any assembler.
const fs=require('node:fs'),crypto=require('node:crypto'),u=require('node:util'),cp=require('node:child_process'),path=require('node:path');
const ROOT='/root/autodl-tmp/symbolic_dynamics',QA=ROOT+'/docs/papers211_215_sequence/qa/';
const HERE=QA+'p211_terminal_binding_source_root01',A=QA+'p211_terminal_binding_independent01';
const ENV={PATH:'/usr/bin:/bin',LANG:'C.UTF-8',LC_ALL:'C.UTF-8',TZ:'UTC'};
const reads={},STAT=['dev','ino','mode','nlink','uid','gid','size','mtimeNs','ctimeNs'];let checks=0;
function need(x,s){checks++;if(!x)throw Error(s);}
function eq(a,b,s){need(u.isDeepStrictEqual(a,b),s);}
function pin(b){return {bytes:b.length,sha256:crypto.createHash('sha256').update(b).digest('hex')};}
function rich(p){
 need(path.isAbsolute(p)&&(p.startsWith(ROOT+'/')||['/usr/bin/node','/usr/bin/cmp'].includes(p)),'exact workspace or named native tool scope');
 const a=fs.lstatSync(p,{bigint:true});need(a.isFile()&&!a.isSymbolicLink()&&fs.realpathSync(p)===p,'ordinary exact path');
 const raw=fs.readFileSync(p),b=fs.lstatSync(p,{bigint:true});const stat=STAT.map(k=>String(a[k]));
 eq(STAT.map(k=>String(b[k])),stat,'stable complete selected stat');
 const key={...pin(raw),stat};need(BigInt(raw.length)===b.size,'whole read size');
 if(reads[p])eq(key,reads[p],'unchanged repeated root original');reads[p]=key;return {raw,key};
}
function read(p){return rich(p).raw;}
function obj(p){return JSON.parse(read(p));}
function put(n,v){const b=Buffer.isBuffer(v)?v:Buffer.from(JSON.stringify(v,null,2)+'\n');const fd=fs.openSync(HERE+'/'+n,'wx',0o600);try{fs.writeFileSync(fd,b);fs.fsyncSync(fd);}finally{fs.closeSync(fd);}}
function seal(base,count,digest){
 const raw=read(base+'/SHA256SUMS');eq(pin(raw).sha256,digest,'exact full independent seal');need(raw.at(-1)===10,'whole manifest LF');
 const rows={},dirs=new Set(['.']);
 for(const line of raw.toString().trimEnd().split('\n')){const m=/^([0-9a-f]{64})  (.+)$/.exec(line);need(!!m,'manifest row');const n=m[2];
  need(!n.startsWith('/')&&!n.split('/').some(x=>!x||x==='.'||x==='..')&&!rows[n]&&n!=='SHA256SUMS','unique safe nonself row');rows[n]=m[1];
  eq(pin(read(base+'/'+n)).sha256,m[1],'whole payload');let par=path.posix.dirname(n);while(par!=='.'){dirs.add(par);par=path.posix.dirname(par);}}
 const files=[],actualDirs=['.'];function walk(d){for(const n of fs.readdirSync(d).sort()){const p=d+'/'+n,s=fs.lstatSync(p);need(!s.isSymbolicLink(),'no alias tree');if(s.isDirectory()){actualDirs.push(path.relative(base,p));walk(p);}else{need(s.isFile(),'ordinary tree file');files.push(path.relative(base,p));}}}walk(base);
 eq(files.sort(),[...Object.keys(rows),'SHA256SUMS'].sort(),'whole file membership');eq(actualDirs.sort(),[...dirs].sort(),'whole directory membership');need(Object.keys(rows).length===count,'whole payload count');return rows;
}
function record(n,label){need(n.request&&typeof n.request.cmd==='string'&&n.result&&n.result.exit_code===0&&!n.result.session_id,'actual settled native '+label);need(typeof n.result.output==='string'&&!n.result.output.includes('tokens truncated'),'complete actual stream '+label);return Buffer.from(n.result.output);}
function command(label,argv,timeout){
 const at={argv,cwd:ROOT,environment:ENV,stdin:'empty pipe',timeout_ms:timeout,started_utc:new Date().toISOString(),executable:rich(argv[0]).key};put(label+'.ATTEMPT.json',at);
 const r=cp.spawnSync(argv[0],argv.slice(1),{cwd:ROOT,env:ENV,input:Buffer.alloc(0),timeout,maxBuffer:4*1024*1024,encoding:null});
 const out=r.stdout||Buffer.alloc(0),err=r.stderr||Buffer.alloc(0);put(label+'.stdout.raw',out);put(label+'.stderr.raw',err);
 const n={...at,ended_utc:new Date().toISOString(),status:r.status,signal:r.signal,error:r.error?{name:r.error.name,message:r.error.message,code:r.error.code||null}:null,stdout:pin(out),stderr:pin(err)};put(label+'.NATIVE.json',n);
 need(r.status===0&&r.signal===null&&!r.error&&err.length===0,'actual native success '+label);eq(rich(argv[0]).key,at.executable,'native tool full unchanged');return out;
}
need(process.cwd()===ROOT,'literal root cwd');eq({...process.env},ENV,'literal ENV4 only');
read(__filename);read(HERE+'/AUTHORITY.md');put('EXECUTED_CAPTURE_SOURCE.js',read(__filename));
seal(A,7,'0efe4f74f47030cef2f2cdbe9c73b2cda50b1f5fd1fe7c2203a15f9fccbc395e');
const source=read(A+'/audit_metadata.js');eq(pin(source),{bytes:20986,sha256:'f68572b2d12ce9a444c4df7784639380da2c68935c7e3e633d8c5cc54d065f44'},'fully read independent source exact');
const n=obj(A+'/NATIVE_CHECK.json'),original=record(n,'independent metadata');eq(pin(original),{bytes:118870,sha256:'471c6feb21760d9b091d5c969f769f6924dfab6a9372e1a8766d344491346664'},'whole original output identity');
const o=JSON.parse(original);need(o.checks===22191&&o.unique_workspace_files===190&&Object.keys(o.read_pins).length===190,'whole original metadata scope');eq(o.read_stat_schema,STAT,'full selected stat schema');
function prior(){for(const [rel,p] of Object.entries(o.read_pins)){need(!rel.startsWith('/')&&rel.split('/').every(x=>x&&x!=='.'&&x!=='..'),'literal original relative');const x=rich(ROOT+'/'+rel);eq(pin(x.raw),p,'all190 original byte pins');eq(x.key.stat,o.read_stats[rel],'all190 original selected stats');}}
prior();put('INPUTS_BEFORE.json',reads);put('INDEPENDENT_ORIGINAL_STDOUT.raw',original);
const actual=command('01_metadata_reuse',['/usr/bin/node',A+'/audit_metadata.js'],60000);eq(actual,original,'entire actual root and independent output raw identical');
eq(command('02_raw_compare',['/usr/bin/cmp','--',HERE+'/INDEPENDENT_ORIGINAL_STDOUT.raw',HERE+'/01_metadata_reuse.stdout.raw'],60000),Buffer.alloc(0),'actual empty raw comparison result');
const close=obj(A+'/CLOSING_NATIVE.json');eq(record(close.source_read,'whole source closing read'),source,'entire actual source read bound');
const c=JSON.parse(record(close.metadata_closing,'metadata closing'));need(c.original_execution_chunk==='cd446b'&&c.input_pins_and_full_selected_stats_rechecked===190&&c.source_lines===244&&c.source_edited_since_execution===false&&c.host_referents_read===0&&c.ambient_environment_read===false,'complete saved independent closure role');
eq(c.original_output_pin,pin(original),'whole saved closing output pin');for(const [name,p] of Object.entries(c.files))eq(pin(read(A+'/'+name)),p,'entire closing originals');
const nr=obj(A+'/NATIVE_READS.json');need(nr.records.length===22,'all22 independent reads');let sourceBindings=0;
for(let i=0;i<nr.records.length;i++){const r=nr.records[i],raw=record(r,'independent read '+i),m=/^sed -n '(\d+),(\d+)p' (\S+)$/.exec(r.request.cmd);if(m&&i!==0){const all=read(ROOT+'/'+m[3]).toString().match(/[^\n]*\n|[^\n]+$/g)||[];eq(raw,Buffer.from(all.slice(Number(m[1])-1,Number(m[2])).join('')),'whole source slice '+i);sourceBindings++;}}
const f=obj(A+'/FINDINGS.json');eq(f.census,{critical:0,major:0,minor:0},'actual zero source finding census');eq(f.findings,[],'actual empty source findings');need(f.verdict==='GO_SOURCE_RECEPTION_ONLY'&&f.operational_status==='HOLD_OPERATIONAL'&&!f.whole_round2_accepted_here&&!f.root_authority_issued_here,'source authority remains separate');
for(const name of ['SOURCE_REPORT.md']){for(const match of read(A+'/'+name).toString().matchAll(/!?\[[^\]\n]*\]\(([^)\n]+)\)/g)){const target=path.resolve(A,match[1].split('#')[0]);need(target.startsWith(ROOT+'/'),'workspace source report link');read(target);}}
prior();seal(A,7,'0efe4f74f47030cef2f2cdbe9c73b2cda50b1f5fd1fe7c2203a15f9fccbc395e');
for(const [p,k] of Object.entries({...reads}))eq(rich(p).key,k,'all root consumed original keys close');put('INPUTS_AFTER.json',reads);
const result={status:'PASS_ROOT_TERMINAL_BINDING_SOURCE_AND_INDEPENDENT_METADATA_RECEPTION_ONLY',checks,read_paths:Object.keys(reads).length,independent_payloads:7,submitted_payloads:o.submitted.payloads,plan_input_pins:166,independent_checks:22191,complete_independent_key:190,independent_native_reads:22,raw_source_read_bindings:sourceBindings,whole_reused_output:pin(actual),native_cmp_exit:0,assembly_executed:false,host_key_refreshed:false,operational_authority:false,round2_accepted_here:false,new_science:0,new_builds:0,new_page_views:0,paper_complete:false,external:'OWNER_AMBER / HOLD_EXTERNAL'};
put('RESULT.json',result);process.stdout.write(JSON.stringify(result)+'\n');
